File: app/src/main/java/com/waveneuro/ui/dashboard/history/HistoryActivity.java
package com.waveneuro.ui.dashboard.history;

import android.os.Bundle;

import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.asif.abase.data.model.BaseModel;
import com.waveneuro.R;
import com.waveneuro.data.model.response.device.SonalDevicesResponse;
import com.waveneuro.ui.base.BaseActivity;

import java.util.ArrayList;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;


public class HistoryActivity extends BaseActivity {
    @Inject
    HistoryViewModel historyViewModel;

    @BindView(R.id.rv_devices_history)
    RecyclerView rvDevicesHistory;
    RecyclerView.Adapter rvDevicesHistoryAdapter;

    private ArrayList<SonalDevicesResponse.Device> devices = new ArrayList<SonalDevicesResponse.Device>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityComponent().inject(this);
        setContentView(R.layout.activity_history);
        ButterKnife.bind(this);

        this.setObserver();

        this.historyViewModel.processEvent(new HistoryViewEvent.Start());

        rvDevicesHistoryAdapter = new DevicesHistoryAdapter(this, this.devices);

        this.rvDevicesHistory.setLayoutManager(new LinearLayoutManager(this));
        this.rvDevicesHistory.setAdapter(rvDevicesHistoryAdapter);
    }

    private void setView() {

    }

    private void setObserver() {
        this.historyViewModel.getData().observe(this, historyViewStateObserver);
    }


    Observer<HistoryViewState> historyViewStateObserver = viewState -> {
        if (viewState instanceof HistoryViewState.Success) {
            HistoryViewState.Success success = (HistoryViewState.Success) viewState;
            onSuccess(success.getData());
        } else if (viewState instanceof HistoryViewState.Failure) {
            HistoryViewState.Failure error = (HistoryViewState.Failure) viewState;
            onFailure(error.getError());
        } else if (viewState instanceof HistoryViewState.Loading) {
            HistoryViewState.Loading loading = ((HistoryViewState.Loading) viewState);
        }
    };

    @Override
    public void onSuccess(BaseModel model) {
        super.onSuccess(model);

        SonalDevicesResponse sonalDevicesResponse = (SonalDevicesResponse) model;

        this.devices.addAll(sonalDevicesResponse.getDevices());

        this.rvDevicesHistoryAdapter.notifyDataSetChanged();
    }

    @OnClick(R.id.iv_back)
    public void OnClickBack() {
        onBackPressed();
    }

}

File: app/src/main/java/com/waveneuro/ui/dashboard/history/HistoryActivity.kt
package com.waveneuro.ui.dashboard.history

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.activity.viewModels
import androidx.lifecycle.Observer
import com.waveneuro.databinding.ActivityHistoryBinding
import com.waveneuro.domain.model.device.Device
import com.waveneuro.ui.base.activity.BaseViewModelActivity
import com.waveneuro.ui.dashboard.history.adapter.DevicesHistoryAdapter
import com.waveneuro.ui.dashboard.history.viewmodel.HistoryViewModel
import com.waveneuro.ui.dashboard.history.viewmodel.HistoryViewModelImpl
import com.waveneuro.utils.ext.getAppComponent

class HistoryActivity : BaseViewModelActivity<ActivityHistoryBinding, HistoryViewModel>() {

    private lateinit var adapter: DevicesHistoryAdapter

    private val devices = mutableListOf<Device>()

    override val viewModel: HistoryViewModelImpl by viewModels {
        getAppComponent().historyViewModelFactory()
    }

    override fun initBinding(): ActivityHistoryBinding =
        ActivityHistoryBinding.inflate(layoutInflater)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setView()
        setObserver()

        viewModel.processEvent(HistoryViewEvent.Start)
    }

    private fun setView() {
        adapter = DevicesHistoryAdapter(this, devices)
        binding.rvDevicesHistory.adapter = adapter
        binding.ivBack.setOnClickListener { onBackPressed() }
    }

    private fun setObserver() {
        viewModel.viewEffect.observe(this, Observer { viewState ->
            when (viewState) {
                is HistoryViewEffect.Success -> {
                    onSuccess(viewState.deviceList)
                }
            }
        })
    }


    private fun onSuccess(deviceList: List<Device>) {
        devices.addAll(deviceList)
        adapter.notifyDataSetChanged()
    }

    companion object {
        fun newIntent(context: Context) = Intent(context, HistoryActivity::class.java)
    }
}